src/Interpreters/MutationsInterpreter.cpp

```diff
@@ -764,4 +764,23 @@ std::optional<SortDescription> MutationsInterpreter::getStorageSortDescriptionIf
     return sort_description;
 }
 
+bool MutationsInterpreter::Stage::isAffectingAllColumns(const Names & storage_columns) const
+{
+    /// is subset
+    for (const auto & storage_column : storage_columns)
+        if (!output_columns.count(storage_column))
+            return false;
+
+    return true;
+}
+
+bool MutationsInterpreter::isAffectingAllColumns() const
+{
+    auto storage_columns = metadata_snapshot->getColumns().getNamesOfPhysical();
+    if (stages.empty())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Mutation interpreter has no stages");
+
+    return stages.back().isAffectingAllColumns(storage_columns);
```

Comment:
Why "back"? Does it mean that if previous stage affect some columns, the next stage also?

Reply:
We can read all columns at some stage, because it may be required by column dependencies. For example, if we want to materialize index or TTL which touches several columns (possibly all). But in this case, we don't need to rewrite them on disk.

Actually output columns of the last stage are columns that will be rewritten.